src/lib/math/pcurves/pcurves_impl.h

```diff
@@ -0,0 +1,1469 @@
+            }
+         }
+
+         m_table = ProjectivePoint::to_affine_batch(table);
+      }
+
+      ProjectivePoint mul(const Scalar& s, RandomNumberGenerator& rng) const {
+         const BlindedScalar bits(s, rng);
+
+         auto accum = [&]() {
+            const size_t w_0 = bits.get_window((Windows - 1) * WindowBits);
+            // Guaranteed because we set the high bit of the randomizer
+            BOTAN_DEBUG_ASSERT(w_0 != 0);
+            auto pt = ProjectivePoint::from_affine(AffinePoint::ct_select(m_table, w_0));
+            pt.ct_poison();
+            pt.randomize_rep(rng);
+            return pt;
+         }();
+
+         for(size_t i = 1; i != Windows; ++i) {
+            accum = accum.dbl_n(WindowBits);
+            const size_t w_i = bits.get_window((Windows - i - 1) * WindowBits);
+
+            /*
+            This point addition cannot be a doubling (except once)
+
+            Consider the sequence of points that are operated on, and specifically
+            their discrete logarithms. We start out at the point at infinity
+            (dlog 0) and then add the initial window which is precisely P*w_0
+
+            We then perform WindowBits doublings, so accum's dlog at the point
+            of the addition in the first iteration of the loop (when i == 1) is
+            at least 2^W * w_0.
+
+            Since we know w_0 > 0, then in every iteration of the loop, accums
+            dlog will always be greater than the dlog of the table element we
+            just looked up (something between 0 and 2^W-1), and thus the
+            addition into accum cannot be a doubling.
+
+            However due to blinding this argument fails, since we perform
+            multiplications using a scalar that is larger than the group
+            order. In this case it's possible that the dlog of accum becomes
+            `order + x` (or, effectively, `x`) and `x` is smaller than 2^W.
+            In this case, a doubling may occur. Future iterations of the loop
+            cannot be doublings by the same argument above. Since the blinding
+            factor is always less than the group order (substantially so),
+            it is not possible for the dlog of accum to overflow a second time.
+            */
+            accum += AffinePoint::ct_select(m_table, w_i);
+
+            if(i <= 3) {
+               accum.randomize_rep(rng);
+            }
+         }
+
+         accum.ct_unpoison();
+         return accum;
+      }
+
+   private:
+      std::vector<AffinePoint> m_table;
+};
+
+/**
+* Effect 2-ary multiplication ie x*G + y*H
+*
+* This is done using a windowed variant of what is usually called
+* Shamir's trick.
+*
+* The W = 1 case is simple; we precompute an extra point GH = G + H,
+* and then examine 1 bit in each of x and y. If one or the other bits
+* are set then add G or H resp. If both bits are set, add GH.
+*/
+template <typename C, size_t W>
+class WindowedMul2Table final {
```

Comment:
Fabian and I had done a refactoring on this, here: https://github.com/randombit/botan/pull/4095/commits/030c919b7fb7ec959cec1a3307e61f66481919de

I don't remember receiving feedback about this commit from you. Just making sure it doesn't fall through the cracks.
At this point it might not apply neatly anymore, though. I'll rebase later.

Reply:
Rebased: https://github.com/randombit/botan/pull/4095/commits/b6e847e4272641d12955e7b64e73a4ad5dc54f73

Reply:
Thanks for the rebase. I did see the commit but still need to think that one over.